### ISSUE TITLE
Auth: OAuth 2.0

### ISSUE THREAD

**Issue by helloanoop** [usebruno/bruno#1003]
Parent Issue: #119

Support OAuth 2.0

**Comment by fuxx** [comment]
Really important feature <3

**Comment by and-rose** [comment]
So far this is the only thing keeping me attached to Postman. Really keen to see this implemented.

**Comment by Rens660** [comment]
We would love to move to Bruno as well, but since we are requiring `grant_type=authorization_code` we can't yet. Is there maybe a progress update on this feature? 🙌

**Comment by tsteckenborn** [comment]
Dependend on what you need you might want to check [this](https://github.com/usebruno/bruno/discussions/385#discussioncomment-7813010) as a workaround for e.g. the Client Credentials flow.

@Rens660 - I've got quite a similar one for Authorization Code. You'll need to do one manual step in the browser, but atleast it returns e.g. the corresponding link as part of an error message. Would you be interested in that?

**Comment by Rens660** [comment]
Hi @tsteckenborn, 
Unfortunately I am stuck with using Authorization Code, but if you would share your workaround for if, I am definitely interested, thanks!

**Comment by helloanoop** [comment]
I will be building OAuth 2.0 functionality in this upcoming live coding stream with Hussain on Dec 16th
https://www.youtube.com/watch?v=ZtNGbglTjps

**Comment by tsteckenborn** [comment]
> Hi @tsteckenborn, Unfortunately I am stuck with using Authorization Code, but if you would share your workaround for if, I am definitely interested, thanks!

Take a look if [that](https://github.com/usebruno/bruno/discussions/385#discussioncomment-7838309) helps to ease the usage

**Comment by mjschlosser8** [comment]
Can't wait for this!

**Comment by andifalk** [comment]
> Pretty much the only thing tying me to insomnia. We use the implicit flow on our end.

FYI: Implicit flow is deprecated, should not be used any more (replaced by Authorization code + PKCE)

**Comment by 42shadow42** [comment]
> > Pretty much the only thing tying me to insomnia. We use the implicit flow on our end.
> 
> FYI: Implicit flow is deprecated, should not be used any more (replaced by Authorization code + PKCE)

Thanks for the heads up, though I don't see Authorization code + PKCE supported here yet either. I'll have to check to see if we have migration plans from Implicit flow.

**Comment by billbrinck** [comment]
Is there any progress on the OAuth2 support?

**Comment by fuxx** [comment]
> Is there any progress on the OAuth2 support?

@helloanoop can you give any news on your progress or estimation when you continue on OAuth 2 support? The past weeks were very quiet around OAuth 2 :cry:

**Comment by helloanoop** [comment]
@fuxx 

Thanks for your patience.

I was a bit tied up in working on the lang improvements that will be needed to unblock some other features in the pipeline. Given the immense pain point and upvotes on this feature, I will spend some time on this week on this ticket and will share some updates towards the end of this week.

**Comment by fuxx** [comment]
Hej @helloanoop :) 

Some weeks passed by and i guess the implementation could be more complicated then expected. Have you gained more insights on how you can proceed?

**Comment by helloanoop** [comment]
Hey @fuxx Appreciate your patience.

A big blocker was the [Bru Lang](https://www.brulang.org/) design which I wanted to complete. This was very important to ensure that we have a way forward to introduce more features in the future and seamlessly allow folks to migrate to the new format (expected to be available by end of Feb)

Now there are two things in my pipeline
1. File Uploads - expected to be launched this week
2. OAuth 2.0 - expected to launched by next week

I am not touching any other feature (except critical bugs) unless above 2 are launched.

**Comment by glitchracer** [comment]
Any news about OAuth 2.0 support?

**Comment by helloanoop** [comment]
We're on the verge of completing OAuth2 integration, just polishing up the last details.
Set to release tomorrow.

![image](https://github.com/usebruno/bruno/assets/9350733/791af098-f7a0-4b32-8313-e50546199521)

**Comment by dougbreaux** [comment]
Hmmm... what's the reason you can't select and set this up at the collection level?

**Comment by tygore587** [comment]
Would also +1 the comment of @dougbreaux. We also have multiple apis where every request if each api uses the same authentication. So it would be cool to have it configured one time per collection (because authentication between apis could be different) and use it for all requests in this collection.

Postman has a good solution here:
You can configure it as `Inherit from parent (which is the collection most of the time)` or `custom for this request` if you have one request in this collection which uses a different method to authenticate.

**Comment by helloanoop** [comment]
OAuth 2.0 is live in `v1.10.0` 🥳 🥳 

![image](https://github.com/usebruno/bruno/assets/9350733/bc338de1-37ed-4f88-93d3-ae95dd21f61e)

**Comment by helloanoop** [comment]
@dougbreaux in the request that fetches a token via OAuth, you should be able to write a post response script like below that saves the access_token to an collection variable. 
```javascript
bru.setVar("oauth_access_token", res.body.access_token)
```
And at the collection level, you could set this token to be sent as a Bearer Auth / In a Header. All request in collections can Inherit this auth by setting the request level auth as `inherit`

@tygore587 @dougbreaux  Wouldn't this work ?

> what's the reason you can't select and set this up at the collection level?

I think this makes sense. We will add support to configure this at the collection level so that there is no need to create a separate request just to do the auth stuff. This could also allow us to implement the refresh token functionality when the token expires.

cc @lohxt1

**Comment by tygore587** [comment]
@helloanoop Thanks for the suggestion. I will check tomorrow at work if this works for me as a work around until you have the feature for the collection.

This is one of the only things it holds me back to suggest it as a postman alternative.

Will the collection wide configuration be scheduled for one of the next releases or are other feature more important to you at the moment? :)

**Comment by helloanoop** [comment]
> Will the collection wide configuration be scheduled for one of the next releases or are other feature more important to you at the moment?

This is priority. We will get this done in the next outgoing release.

> I will check tomorrow at work if this works for me as a work around until you have the feature for the collection.

Thanks! If at all you need help, I'd be happy to get on a call and work through this. I am available 9am-6pm GMT. My email id is anoop@usebruno.com (if you want to send an invite)

Above invitation ⬆️ applies to everyone. I would love to get on a call and help you through any issues you are facing. All of this is free consultation :) Email me at anoop@usebruno.com and I'd be happy to get on 1:1 call.

**Comment by lhanson** [comment]
I'll second @tygore587 's sentiment, I require OAuth 2.0 for APIs I work on, can't use Postman on NixOS, and refuse to waste my time on the workarounds due to their user-hostile practice of removing older versions from the internet when they do releases.

Much respect to @helloanoop for being proactive and receptive to user needs, once collection-based config lands I'll definitely be recommending Bruno to my team.

**Comment by timozn** [comment]
Thank you for implementing this.
The fact that you have this implemented this with having the option for a Callback URL makes Bruno stand out of it's competitors, stepping in for Postman, quite a lot.

However I am missing a way to provide a Code Challenge Method (like SHA-265 or Plain) for the PCKE, as we need to provide `code_challenge` like defined in [RFC7636](https://www.rfc-editor.org/rfc/rfc7636)

**Comment by tygore587** [comment]
@helloanoop Again thanks for implementing it. I just tested it and want to give some feedback.

First like @timozn says, we need the grant type `Authorization code (with PKCE)`

I prepared a working example in postman.

I used https://demo.duendesoftware.com/ for it to test bruno and postman.

My configuration:
Grant Type: `AuthorizationCode (With PKCE)`
Callback URL: `https://demo.duendesoftware.com/`
Auth URL: `https://demo.duendesoftware.com/connect/authorize`
Access Token URL: `https://demo.duendesoftware.com/connect/token`
Client ID: `interactive.confidential`
Client secret: `secret`
Code Challenge method: `SHA-256`
ode Verifier: empty
Scope: `openid profile email api offline_access`
State: empty

If you don't want to configure it yourself, I prepared a [Postman collection](https://github.com/usebruno/bruno/files/14432607/Dudent.Auth.test.postman_collection.json) where I configured auth on the collection level and also a 
[Bru Request file](https://github.com/usebruno/bruno/files/14432651/Auth.Request.bru.txt)(I needed to add a `.txt` at the end, because otherwise I could not upload the file) for you.

In Postman it opens a browser where I can login with the `alice/alice` demo account and I get back my token.

In bruno I get:
![image](https://github.com/usebruno/bruno/assets/12038641/7dcc6cc8-2ec0-42ce-a531-e78cff836691)

I hope this helps. If you need any further information, please let me know :)

**Comment by ThomasWendrock** [comment]
> @helloanoop Again thanks for implementing it. I just tested it and want to give some feedback.
> 
> First like @timozn says, we need the grant type `Authorization code (with PKCE)`
> 
> I prepared a working example in postman.
> 
> I used https://demo.duendesoftware.com/ for it to test bruno and postman.
> 
> My configuration: Grant Type: `AuthorizationCode (With PKCE)` Callback URL: `https://demo.duendesoftware.com/` Auth URL: `https://demo.duendesoftware.com/connect/authorize` Access Token URL: `https://demo.duendesoftware.com/connect/token` Client ID: `interactive.confidential` Client secret: `secret` Code Challenge method: `SHA-256` ode Verifier: empty Scope: `openid profile email api offline_access` State: empty
> 
> If you don't want to configure it yourself, I prepared a [Postman collection](https://github.com/usebruno/bruno/files/14432607/Dudent.Auth.test.postman_collection.json) where I configured auth on the collection level and also a [Bru Request file](https://github.com/usebruno/bruno/files/14432651/Auth.Request.bru.txt)(I needed to add a `.txt` at the end, because otherwise I could not upload the file) for you.
> 
> In Postman it opens a browser where I can login with the `alice/alice` demo account and I get back my token.
> 
> In bruno I get: ![image](https://private-user-images.githubusercontent.com/12038641/308514960-7dcc6cc8-2ec0-42ce-a531-e78cff836691.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMTczNjYsIm5iZiI6MTcwOTExNzA2NiwicGF0aCI6Ii8xMjAzODY0MS8zMDg1MTQ5NjAtN2RjYzZjYzgtMmVjMC00MmNlLWE1MzEtZTc4Y2ZmODM2NjkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI4VDEwNDQyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkYjIyMWNiZDU3ODJlZjFhYWE5ZTUyYTRlZTZkNzU2ZjBiYjZkMzRjY2I1OGFhZWRhMDlmMGZhZmQ3MzM5NzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QuA0oeOu6WO65qGzhQK8Ao8RxULuuIfm9c5lpqqdX2Q)
> 
> I hope this helps. If you need any further information, please let me know

I have also an error with the Authorization code (flow).
![image](https://github.com/usebruno/bruno/assets/46812473/3b6b4369-2988-44ac-96f1-871850fc1108)

**Comment by Schwobaland** [comment]
> Hi @helloanoop, first of all thanks for your implementation. Could be that in grant type client_credentials you are missing an option to set the access token url?

Also, thank you very much for implementing this. This has been a highly anticipated feature.

For client credentials both access token url and and also the scope is missing for us.

**Comment by vlaurin** [comment]
@helloanoop Thank you very much for your work on this.

A tiny bit of feedback, though of low priority:

* **Infinitely stackable authorization windows**: Currently a new user authorization window opens every time the "Get Access Token" button is clicked which may lead to many authorization windows being super-imposed. I don't think it makes sense to have more than one open at a time so it could be worth considering aborting and closing any previous authorization window when the "Get Access Token" button is clicked again.
* **Error handling on redirect**: As per [OAuth spec](https://www.oauth.com/oauth2-servers/server-side-apps/possible-errors/), if the authorization fails for some reason the IdP will redirect with a parameter `error` instead of `code` (and optionally `error_description`/`error_uri`). As it stands in Bruno the authorization window remain open and doesn't intercept the redirect. This can make errors hard to identify. It would be nice if the error was intercepted, the authorization window closed and the error displayed in Bruno's main window.

**Comment by jackj-msft** [comment]
@helloanoop - OAuth2.0 expects URL encoded form instead of JSON content
https://www.oauth.com/oauth2-servers/server-side-apps/example-flow/

Sample request body from official documentation
```
code=Yzk5ZDczMzRlNDEwY
&grant_type=code
&redirect_uri=https://example-app.com/cb
&client_id=mRkZGFjM
&client_secret=ZGVmMjMz
&code_verifier=Th7UHJdLswIYQxwSg29DbK1a_d9o41uNMTRmuH0PM8zyoMAQ
```

This is why @ThomasWendrock is getting the missing `grant_type`

**Comment by AndreasBAtT** [comment]
How to set up OAuth 2 with Bearer Token ?

**Comment by wtrocki** [comment]
> How to set up OAuth 2 with Bearer Token ?

That would not require OAuth. Any bearer token can be sent using "Bearer Token" authentication

**Comment by AndreasBAtT** [comment]
But Tools like Insomnia use Oauth2 to get Bearer token and automatically set then the bearer token for the API Request. That would be great to have also in Bruno. How is it possible to do so?

**Comment by brki** [comment]
> But Tools like Insomnia use Oauth2 to get Bearer token and automatically set then the bearer token for the API Request. That would be great to have also in Bruno. How is it possible to do so?

@AndreasBAtT It is possible to do it in two steps, e.g. 1) get the auth token and set the corresponding env variable, and then 2) use the Bearer token in your other request(s).

See the videos at https://github.com/usebruno/bruno/pull/1654#issue-2150484049 .

**Comment by wtrocki** [comment]
@rmueller83 I believe that this was fixed in master version. Please check what version you are using

**Comment by rmueller83** [comment]
@wtrocki what do you mean with main version? As I wrote above I am using the most recent version 1.11.0

**Comment by wtrocki** [comment]
@rmueller83 You can view raw request/response in the Timeline Tab. It looks like there still might be the issue with request content type.

**Comment by rmueller83** [comment]
I first configured OAuth2 on the collection, and there is no timeline tab.
Then I configured it for a request and in the Timeline I see a GET request for the "real" REST endpoint, but not for the OAuth2 endpoint, although the JSON data in the second line contains the OAuth payload. Is this only a display error?
The response is 405, but I am not sure against which endpoint the request went...

**Comment by michaelbrehm** [comment]
Using Bruno 1.11.0 with OAuth 2.0 Authorization Code against Okta and I'm getting an error about invalid 'state' parameter.
![image](https://github.com/usebruno/bruno/assets/599117/a2ed5c1e-8b30-45ad-92f2-26f4ae3a0581)

Attempting to move from Insomnium and it has a section "Advanced Options" that contains the Scope, State as well as other options that I don't use.

**Comment by tygore587** [comment]
@lohit-1 @helloanoop another feedback for 1.11.0: I tried oAuth2.0 with pkce again and using [this](https://github.com/usebruno/bruno/issues/1003#issuecomment-1968704251) configuration from the comment above but the only thing I get back is as a response:

<img width="949" alt="image" src="https://github.com/usebruno/bruno/assets/12038641/8f02ba13-67ba-452b-9e5f-510e62aee009">

**Comment by meldron** [comment]
In version `v1.11.0` I can use the `Authorization Code` flow to get a `access_token`, but if I want to do the request to the "normal" target afterwards, it's just always redoing the Oauth2 Request. What do I have to do to set the access token as `Authorization: Bearer` header after the response of `Get Access Token` was received?

**Comment by bacheric** [comment]
> Hmmm... "Cannot inherit Oauth2 from collection" ![image](https://private-user-images.githubusercontent.com/25640850/312243915-fd13fd9e-66b2-4c71-b9cf-996ed38c646e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2MjM3MDgsIm5iZiI6MTcxMDYyMzQwOCwicGF0aCI6Ii8yNTY0MDg1MC8zMTIyNDM5MTUtZmQxM2ZkOWUtNjZiMi00YzcxLWI5Y2YtOTk2ZWQzOGM2NDZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE2VDIxMTAwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwM2U2NDlhYjJiNTBjYjM2N2U4OTM3ZjBkZWRkZjk2ZTgwZGE2NmMxMDgxYWEzODhhY2VlNDczZmE1MDQ2Y2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FXdTlDFnXLha_Lb4hA4DWQZ5PIta0pYjbajsThWQq8k)

Is this a bug or working as intended?

**Comment by MarkLeMerise** [comment]
It's disappointing that OAuth2 inheritance is not yet working, but I assume that's coming as it would get tedious quickly to copy-paste tokens into the request headers.

That aside, Not sure if I should open a whole new issue for this, but the OAuth2 client credentials flow is not working for me at all. I have the equivalent flow working in P***man and have quadrupled checked the settings: URL, client ID, secret, and scope. As a sanity check, I was able to make the equivalent request myself and get a token:

![Screenshot 2024-03-18 at 10 15 24 AM](https://github.com/usebruno/bruno/assets/573634/7408345d-26db-401f-9b21-bcdc27adb88c)

When I try using the collection-level OAuth2 settings, I get an "invalid_request" message.

![Screenshot 2024-03-18 at 10 16 10 AM (2)](https://github.com/usebruno/bruno/assets/573634/f819f3fb-0655-41f2-b853-2f0781a85d2e)

That being said, while this is the primary blocking issue stopping me from switching from my aforementioned client, I think Bruno has great potential and I love seeing the API client space getting disrupted. Thanks for all the hard work!

**Comment by layereight** [comment]
OAuth 2.0 Authorization Code Flow fails for me when the authorization server does a specific intermediate redirect. It's an edge case but the fix is very simple:

https://github.com/usebruno/bruno/pull/1777

A corresponding bug ticket exists: https://github.com/usebruno/bruno/issues/1778

The issue still exists with the new release 1.11.0

**Comment by francoischaix-circeo** [comment]
Same issue as @MarkLeMerise a `client_credential` flow ends in invalid request although data is correct when used in collection or in request directly.

That being said, there is some configuration missing:

- Where should the authorization data be added to:
-- Either in request headers
-- as a query param in the request url (RFC 6750 2.3 URI Query param) 

In case the of header, what is the prefix to add (Bearer in most cases).
In case or query param, the name of the query param

**Comment by jackj93** [comment]
> It's disappointing that OAuth2 inheritance is not yet working, but I assume that's coming as it would get tedious quickly to copy-paste tokens into the request headers.
> 
> That aside, Not sure if I should open a whole new issue for this, but the OAuth2 client credentials flow is not working for me at all. I have the equivalent flow working in P***man and have quadrupled checked the settings: URL, client ID, secret, and scope. As a sanity check, I was able to make the equivalent request myself and get a token:
> 
> ![Screenshot 2024-03-18 at 10 15 24 AM](https://private-user-images.githubusercontent.com/573634/313713089-7408345d-26db-401f-9b21-bcdc27adb88c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA3NzY3NTYsIm5iZiI6MTcxMDc3NjQ1NiwicGF0aCI6Ii81NzM2MzQvMzEzNzEzMDg5LTc0MDgzNDVkLTI2ZGItNDAxZi05YjIxLWJjZGMyN2FkYjg4Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxOFQxNTQwNTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MTM5NmJlNjQzYmQzZDFmNDQ5NzYzNTU1ZWMwN2ViMGZkOThkOWYwZTVkZmQyYTQ5NmEzMmYwMGY4NmE3YTQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.b7SYmy0RrFdggOjdFIWMi6ONtRz0rIfQtNKOeESnBL4)
> 
> When I try using the collection-level OAuth2 settings, I get an "invalid_request" message.
> 
> ![Screenshot 2024-03-18 at 10 16 10 AM (2)](https://private-user-images.githubusercontent.com/573634/313713237-f819f3fb-0655-41f2-b853-2f0781a85d2e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA3NzY3NTYsIm5iZiI6MTcxMDc3NjQ1NiwicGF0aCI6Ii81NzM2MzQvMzEzNzEzMjM3LWY4MTlmM2ZiLTA2NTUtNDFmMi1iODUzLTJmMDc4MWE4NWQyZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxOFQxNTQwNTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00OTA4OTA3MmRiNWQwMjBlOThiZDk5ZTRiZjU1YTI1NzM5ZWE0YzFjODY4NzdiOGQ1NmU5OGI4MWJjMzg3MGM4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.RT1uU8J1_kFWti9Q6-lBde6HDVDCbYwBYeOOwWRtAdM)
> 
> That being said, while this is the primary blocking issue stopping me from switching from my aforementioned client, I think Bruno has great potential and I love seeing the API client space getting disrupted. Thanks for all the hard work!

There is a workaround for to get collection level OAuth working here https://github.com/usebruno/bruno/issues/1704#issuecomment-2002763091

**Comment by dougbreaux** [comment]
For me, OAuth continues to not work at all, either at the Collection level or at the individual request level. And with there being no way to see the "Get Access Token" request/response, I can't even troubleshoot. 

Instead, I have to resort to my prior pattern of creating a specific request to get the token and script that to save off the token into a variable, where all my other requests put that token in an Authorization Bearer header.

(In fact, _none_ of the Collection level authorizations have worked for me at all, including Basic Auth)

**Comment by francoischaix-circeo** [comment]
The OAuth has several issues.
- Grant type is not sent resulting in invalid request
- If the request is a get then a get is sent to get the token instead of a POST

As Axios is a library included I would actually write the connexion script directly in the pre-request script....

**Comment by palhal** [comment]
I can't believe OAuth2 support has been released when it's so poorly implemented. These are the issues I've found after two minutes of testing with version 1.12.2:

1. OAuth2 payload is sent as JSON instead of form data, resulting in _unsupported_grant_type_
2. If the request is GET, the token request is also sent as GET, resulting in 405 Method Not Allowed
4. Not possible to see raw auth request at collection level
5. Data shown in the request timeline seems unreliable (OAuth2 payload sent to original request URL)
6. Client ID and Client Secret are missing from Password Credentials flow
8. Scope is sent as "" when blank (should be left out?)
9. Not possible to inherit Auth from collection

**Comment by ThomasPilot** [comment]
> Thanks for the awsome work regarding OAuth 2.0 Authorization Code Flow. Unfortunately, it fails for me when the authorization server does a specific intermediate redirect. It's an edge case but the fix is very simple:
> 
> See my PR: #1777
> 
> A corresponding bug ticket exists: #1778
> 
> The issue still exists with the new release 1.11.0

I have the same issue with redirects in an OAuth flow. Accepting the linked PR would really help :)

**Comment by pascalknupper** [comment]
Is there any update on this topic, can´t wait to see OAuth 2 auth flow without scripting in Bruno 🙂

**Comment by StefanTUI** [comment]
Bruno is great! And I'd also love to see the OAuth2 auth flow (grant type client credentials) completed, so I fully recommend it everywhere I can :-)

**Comment by helloanoop** [comment]
Hey everyone,

Thanks for all the feedback around improvements on the OAuth2 implementation. We know this has been on the wishlist for a while, and we’re finally diving into a full revamp of the OAuth2 flow. We’re aiming to wrap this up by end of December and will keep you posted here as we make progress.

As part of this process, we’re forming a cohort to better understand how you use OAuth2 in Bruno and gather insights into your expectations. If you’d like to join, please fill out this form: https://forms.gle/gicLNdt5F1XD1MBq7
In the discussions, we'd like to know and understand
* How you’re currently using OAuth2 in Bruno.
* Any challenges or limitations you’ve faced with the current implementation.
* Features or improvements you’d like to see in the revamped version.

We are also reviewing the PRs - #2164, #2077, #2058 as a part of this effort.

Thanks,
Bruno Team

**Comment by JoniJnm** [comment]
How can I fill out the form? 

I only wanted to say: I would like to be able to open an external browser during the oauth2 `authorization_code` to use the local cookies and do not need to login every time in the external provider. Thanks.

![image](https://github.com/user-attachments/assets/a3f32f04-9d9e-4d60-93d3-dc5dfcca3c19)

**Comment by helloanoop** [comment]
There was a permission issue in the cohort participation form. This has been fixed now.

**Comment by tygore587** [comment]
> How can I fill out the form?
> 
> I only wanted to say: I don't know if it's possible, but I would like to be able to open an external browser (instead of in-app) during the oauth2 `authorization_code` to use the local cookies and do not need to login every time in the external provider. Thanks.
> 
> ![image](https://private-user-images.githubusercontent.com/577282/390799820-a3f32f04-9d9e-4d60-93d3-dc5dfcca3c19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4MjA5MDgsIm5iZiI6MTczMjgyMDYwOCwicGF0aCI6Ii81NzcyODIvMzkwNzk5ODIwLWEzZjMyZjA0LTlkOWUtNGQ2MC05M2QzLWRjNWRmY2NhM2MxOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOFQxOTAzMjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02M2QxZjc4MjQ1YzdjZTgwYjIwOWU3NzJhZDkyYmMxNzE2NTk3YjExNWQzM2JiNzEzMzgwZTk4NWRjMzY0YzU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.YZtFPMG6fyp5AhKbp8iH9hOaHjevbxN2qktE1xi6Zwg)

Also a +1. We need to login with a device id in entra id. This device id is only send by edge or chrome with an extension and for this we need to login via an external browser.

**Comment by caviyacht** [comment]
I hope this comes out soon, I want to use Bruno (looks amazing) but am 100% reliant on the implicit grant flow. Any update as we're now in January?

**Comment by shonigbaum** [comment]
When is the OAuth 2.0 coming? Still getting a "invalid request" message when trying it:

![Image](https://github.com/user-attachments/assets/11119641-0bd7-4c06-b2da-f6616652341e)

**Comment by bigsing** [comment]
@shonigbaum Per the Bruno Discord posted on 1/28:

Some polish is pending.
Here is a experimental build if you want to play around https://github.com/usebruno/bruno-experimental-builds/releases/tag/oauth2-revamp-v1.39.0-2025.1.24

Core functionality is done. We need to provide a setting to automatically fetch the token.
Let us know if you encounter any issues in the above build and if you have feedback on what else needs to be improved.

hoping to put out a public beta during feb first week and then release it live on feb 3rd week.

**Comment by abieler-sap** [comment]
I tried the preview and got an error fetching a token. Is there a possibility to analyze what went wrong?
Is there a plan to add the token request also to the timeline, so a user can see what's happening in the token request?

**Comment by T-fala** [comment]
The Experimental build (1.39.0) fixed my usecases (client-credential flow) - Thanks! 🥇

**Comment by anusree-bruno** [comment]
> I tried the experimental build and got an error fetching a token. Is there a possibility to analyze what went wrong? Is there a plan to add the token request also to the timeline, so a user can see what's happening in the token request?

We're working on a console/timeline feature to display all redirects that occur during any request.

**Comment by anusree-bruno** [comment]
> The Experimental build (1.39.0) fixed my usecases (client-credential flow) - Thanks! 🥇
> 
> Edit: Will there be an Option to refetch expired tokens?

Yes, we're actively working on this now!

**Comment by abieler-sap** [comment]
I just tested the newest preview version.
https://github.com/usebruno/bruno-experimental-builds/releases/tag/oauth2-revamp-v1.39.0-2025.2.11

For our use case the build in OAuth authentication seems to work now to get a token and use it.
Additionaly it would be great, if the token would be fetched automatically. Of course it should be cached until it expires ;-)

This is really a great step forward. Looking forward to see the improvements in the official versions 👍

**Comment by anusree-bruno** [comment]
@abieler-sap Thanks for the feedback!

We're adding an option to automatically fetch tokens, which can be enabled or disabled based on user preference.

**Comment by tygore587** [comment]
@anusree-bruno It would be nice to have an option to use an external browser instead of the one build in to bruno. I need the browser to send a device id, so I can obtain a token.

**Comment by anusree-bruno** [comment]
Hi @tygore587,

I know this is a highly requested feature. We're working on it, and it will be available in the next phase of the OAuth2 updates.

**Comment by Manimu** [comment]
Hi!

I also tested the preview version from https://github.com/usebruno/bruno-experimental-builds/releases/tag/oauth2-revamp-v1.39.0-2025.2.11

I really like how the new OAuth wizard looks like! Specifically I like how the token with expiration is now displayed and how I am now able to configure the token properly (e.g. header prefix). Great work!! :thumbsup:

However, I noticed two things:

1. In the non-preview version I can see the response of the token request as json in the Response tab. In the preview version this is no longer the case. Is this intentional? I am not sure, if I will miss it, but it was nice to see the raw response when debugging issues.
2. The OAuth wizard does not work in my company network behind a proxy. This is true for both: the current version and the preview version of Bruno. I am not sure if this is a Bruno issue or not. But my old pinned Postman version has no issues behind a proxy.

**Comment by anusree-bruno** [comment]
Hi @Manimu,

Thanks for your feedback! Really glad to hear you’re liking the new improvements we’ve made! 😊

1. The response of the token request is no longer visible in the Response tab, as that space is now dedicated to displaying the resource URL response. However, we plan to log all requests and responses made during the authentication process in a console to maintain visibility for debugging.

2. Regarding the proxy issue, have you tested this on the latest build? We released an update a few days ago with fixes related to proxy and TLS issues. Please let me know if you're still facing the problem, and we can investigate further.

**Comment by dr-schopalopp** [comment]
any plans when the revamp will make it into a stable release?

the prerelease `v1.39.0-2025.2.11` is exactly what I'm waiting for

**Comment by Manimu** [comment]
Additional info regarding my proxy issue: Everything works when I configure my proxy manually (Proxy Mode "On") instead of using environment variables (Proxy Mode "System Proxy"). So for now I am just using the first mode and the issue is resolve for me. 

Thanks!

**Comment by MikePhub** [comment]
feature request: it would be nice if it could store multiple authorizations (access/refresh tokens) so i can toggle between tokens when testing in the same collection without needing to repeat the oauth flow each time i switch users.

**Comment by BBE-91** [comment]
Oauth2 implicit has just been released, I was expecting this since a long time
However, I can't use so far 🥺 as I got an error while fetching the token:  invalid response type
The reason is the response_type is forced to "token"
Is there anywhere a setup to change it for "id_token" or both "id_token token" ?
Thanks in advance

**Comment by pietrygamat** [comment]
@BBE-91 the [RFC](https://datatracker.ietf.org/doc/html/rfc6749#section-4.2.1) mandates that the authorization request expects response_type to be "token" - if your server expects anything else - it's not compliant with OAuth2, and that's quite wierd.

**Comment by anusree-bruno** [comment]
Closing this umbrella issue. The core OAuth 2.0 work tracked here has shipped and the remaining items are now tracked in dedicated issues.

If you still hit an OAuth 2.0 problem, please open a new, focused issue with steps to reproduce and your Bruno version.

Related follow-ups we’re tracking separately:

- https://github.com/usebruno/bruno/issues/2002
- https://github.com/usebruno/bruno/issues/2650

Thanks everyone who reported, tested, and contributed! 🙏
